src/Avalonia.Controls/TextBlock.cs

```diff
@@ -811,6 +811,7 @@ protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs chang
                 case nameof(TextDecorations):
                 case nameof(FontFeatures):
                 case nameof(Foreground):
+                case nameof(Bounds):
```

Comment:
I don't see why we ever wanted to listen to Bounds changes. The parent is responsible for invalidating its children.

Reply:
if Bounds was the underlaying issues, this test should fail, no? at least the one inside the Border. However, all of it works as intended: 

```xml
<StackPanel>
    <Slider x:Name="Sld" Minimum="100" Maximum="500"></Slider>
    <TextBlock Text="Test" Width="{Binding #Sld.Value}" Background="LightGreen" TextAlignment="Right"></TextBlock>
    <Border Width="{Binding #Sld.Value}" Background="Yellow">
        <TextBlock Text="Test" TextAlignment="Right" />
    </Border>
    <TextBox Text="Test" Width="{Binding #Sld.Value}" TextAlignment="Right"></TextBox>
</StackPanel>
```

... So I think @Gillibald is right, the issue may be somewhere inside Grid.

Reply:
The logic you are advocating is that the measurement of TextLayout should remain valid after the width of TextBox changes. This seems obviously not true, especially in cases of justification other than Left which need to grow the size of TextLayout to add the necessary white space.  Why wouldn't TextBox be responsible for invalidating its own children?

I didn't look at Slider, but it could be dong aggressive invalidation that Grid can not get by with.  Grid will quickly end up in a layout loop with too much invalidating.

With Grid, there is only one Measure pass for each child starting from (Infinity, Infinity).  At this point in my research, I believe that Grid assumes that children won't return a different measurement of themselves during future Arrange passes when they were provided with as much space as needed initially.

In this case, there are TWO independent grids, with certain properties being kept in sync by DefinitionBase in an OnLayoutUpdated method.  Invalidating the other grid there can quickly start a layout loop if done at the wrong time.

During a subsequent Arrange pass, Grid B Column A updates the minimum width of Grid A Column A, which causes another Arrange pass. The TextBlock is passed a larger size and asked to Arrange with its measure still valid.  This changes Bounds in the Layoutable base class in ArrangeCore, which has no knowledge of who has subclassed it, or of the interior TextLayout component.

A change could be made to Grid to invalidate ALL cell contents completely, but this would cost more of a performance hit and more invalidation in cells that do not contain a TextBlock than having only TextBlock cells selectively invalidate their own TextLayout area only in columns which have changed their width.

Reply:
The Slider in your example is bound to the Layoutable.Width property of the text controls, not their Visual.Bounds property.  That is the difference between how the TextBlocks in bug #13777 are updated, and how your example is set up.  (It is not possible to set up a test using Bounds because it is a readonly property.)

Layoutable has a series of AffectsMeasure properties, _but Bounds is not one of them_, even though changes to Bounds affects Width and Height!  TextBlock _invalidates itself_ in the AffectsMeasure static method - it is not done by the Slider - when any of the following properties are changed:

 ```
static Layoutable()
        {
            AffectsMeasure<Layoutable>(
                WidthProperty,
                HeightProperty,
                MinWidthProperty,
                MaxWidthProperty,
                MinHeightProperty,
                MaxHeightProperty,
                MarginProperty,
                HorizontalAlignmentProperty,
                VerticalAlignmentProperty);
        }
```

Reply:
I thought the same before, that's why another TextBlock is inside a Border element which should indirectly change bounds. We may want to change Border to Panel or whatever but I still think it should behave the similar as SharedGridSize does. However it doesn't, at least on my side. 🤷‍♀️

Reply:
The difference with the TextBlock in a Border is that Border calls TextBlock.Measure with a _constrained_ width of (Border.Width, Infinity) instead of (Infinity, Infinity).  

The Grid always measures children in auto width columns with (Infinity, Infinity) per this comment:

```
 1101 //  if cell belongs to at least one Auto column and not a single Star column
 1102 //  then it should be calculated "to content", thus it is possible to "shortcut"
 1103 //  calculations and simply assign PositiveInfinity here.
 1104 cellMeasureWidth = double.PositiveInfinity;

```

Bounds changes during Measure in your example.
Bounds changes during Arrange with SharedGridSize.

Reply:
This check is probably wrong: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBlock.cs#L736
The arrange call on TextBlock should still invalidate the TextLayout if the finalSize differs from the previous measure.

Reply:
```diff
            var padding = LayoutHelper.RoundLayoutThickness(Padding, scale, scale);

            //Fixes: #11019
-            if (finalSize.Width < _constraint.Width)
+            if (!MathUtilities.AreClose(finalSize.Width, _constraint.Width))
            {
                _textLayout?.Dispose();
                _textLayout = null;
```
I have updated the PR to invalidate on unequal width at line 736 instead of watching the Bounds property.  I think it would be a good idea to proactively do the same check on the Height property.  Does anyone agree?